### PR TITLE
[InstCombine] Simplify nested selects with implied condition

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch does the following simplification:
```
sel1 = select cond1, X, Y 
sel2 = select cond2, sel1, Z
-->
sel2 = select cond2, X, Z if cond2 implies cond1
sel2 = select cond2, Y, Z if cond2 implies !cond1
```
Alive2: https://alive2.llvm.org/ce/z/9A_arU

It cannot be done in CVP/SCCP since we should guarantee that `cond2` is not an undef.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 1]
```diff
@@ -3867,5 +3867,22 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
...
+    if (auto Implied =
+            isImpliedCondition(CondVal, InnerCondVal, DL, /*LHSIsTrue=*/true))
+      return replaceOperand(SI, 1, *Implied ? LHS : RHS);
+  if (match(FalseVal,
+            m_Select(m_Value(InnerCondVal), m_Value(LHS), m_Value(RHS))) &&
+      CondVal->getType() == InnerCondVal->getType())
+    if (auto Implied =
+            isImpliedCondition(CondVal, InnerCondVal, DL, /*LHSIsTrue=*/false))
+      return replaceOperand(SI, 2, *Implied ? LHS : RHS);
+
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3886]: Maybe put this in a lambda:
```
auto CanSimplfySelectArmAs = [...](bool Arm) {
 if(match(Arm ? TrueVal : FalseVal, m_Select(...)) && CondVal->getType() == ...)
  if(Implied = isImplied(..., Arm)
   return *Implied ? LHS : RHS;
}
if(auto * V = CanSimplfySelectArmAs(true)) replaceOperand(SI, 1, V);
if(auto * V = CanSimplfySelectArmAs(false)) replaceOperand(SI, 2, V);

```
?

There is a lot of duplication between the two arms.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3889]: Fixed.

**Review by nikic — COMMENTED** [review]
This looks basically fine, but I think it may subsume some existing folds?

For example, what about this one? https://github.com/llvm/llvm-project/blob/27ce5121ee875c337dd0f977b00afcb756977f62/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp#L3646-L3651

There's also foldAndOrOfSelectUsingImpliedCond() -- I think your fold may be the same, but without the limitation that the top-level select is a logical one?

**Comment by dtcxzyw** [comment]
> This looks basically fine, but I think it may subsume some existing folds?
> 
> For example, what about this one?
> 
> https://github.com/llvm/llvm-project/blob/27ce5121ee875c337dd0f977b00afcb756977f62/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp#L3646-L3651
> 
> There's also foldAndOrOfSelectUsingImpliedCond() -- I think your fold may be the same, but without the limitation that the top-level select is a logical one?

We can further remove the following logic if we fold it recursively.
https://github.com/llvm/llvm-project/blob/d0521484c854adf082936824c99f9db101363cb8/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp#L3682-L3723

**Comment by dtcxzyw** [comment]
> This looks basically fine, but I think it may subsume some existing folds?
> 
> For example, what about this one?
> 
> https://github.com/llvm/llvm-project/blob/27ce5121ee875c337dd0f977b00afcb756977f62/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp#L3646-L3651
> 
> There's also foldAndOrOfSelectUsingImpliedCond() -- I think your fold may be the same, but without the limitation that the top-level select is a logical one?

Done.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
> This looks basically fine, but I think it may subsume some existing folds?
> 
> For example, what about this one?
> 
> https://github.com/llvm/llvm-project/blob/27ce5121ee875c337dd0f977b00afcb756977f62/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp#L3646-L3651
> 
> There's also foldAndOrOfSelectUsingImpliedCond() -- I think your fold may be the same, but without the limitation that the top-level select is a logical one?

The compile-time improvement looks good:
http://llvm-compile-time-tracker.com/compare.php?from=732a5cba8c739ed40a7280b5d74ca717910c2c4c&to=2e94109d115a81f271cefc7d6c0e839920c52bba&stat=instructions%3Au

**Pull request merged**